File: skills/cumcm-code-writer/references/algorithm_library/01_优化与启发式/nsga2/matlab/main.m
% NSGA-II · 双目标优化（MATLAB 最小可运行模板）
% 模型：min f1 = sum(x.^2), min f2 = sum((x-2).^2)，x∈[0,2]，n=10
here = fileparts(mfilename('fullpath'));
addpath(fullfile(here, '..', '..', '..', '_common', 'matlab'));

dims = 10; low = 0; high = 2; popSize = 40; generations = 30;
obj = @(P) [sum(P.^2, 2), sum((P - 2).^2, 2)];
rng(42);
pop = low + (high - low) * rand(popSize, dims);
objs = obj(pop);

for g = 1:generations
    [fronts, rank] = ndsort(objs);
    crowd = zeros(popSize, 1);
    for f = 1:numel(fronts)
        crowd(fronts{f}) = crowdDist(objs, fronts{f});
    end
    children = zeros(popSize, dims);
    for k = 1:2:popSize
        a1 = tournament(rank, crowd);
        a2 = tournament(rank, crowd);
        u = rand(1, dims);
        beta = ((2 * u) .^ (1/3)) .* (u <= 0.5) + ((1 ./ (2 * (1 - u))) .^ (1/3)) .* (u > 0.5);
        c1 = 0.5 * ((1 + beta) .* pop(a1, :) + (1 - beta) .* pop(a2, :));
        c2 = 0.5 * ((1 - beta) .* pop(a1, :) + (1 + beta) .* pop(a2, :));
        mut1 = rand(1, dims) < 0.1; c1(mut1) = c1(mut1) + 0.1 * randn(1, sum(mut1));
        mut2 = rand(1, dims) < 0.1; c2(mut2) = c2(mut2) + 0.1 * randn(1, sum(mut2));
        children(k, :) = min(max(c1, low), high);
        if k + 1 <= popSize
            children(k + 1, :) = min(max(c2, low), high);
        end
    end
    merged = [pop; children];
    mergedObj = obj(merged);
    [mf, ~] = ndsort(mergedObj);
    newPop = []; newObj = [];
    for f = 1:numel(mf)
        idx = mf{f};
        if size(newPop, 1) + numel(idx) <= popSize
            newPop = [newPop; merged(idx, :)]; %#ok<AGROW>
            newObj = [newObj; mergedObj(idx, :)]; %#ok<AGROW>
        else
            d = crowdDist(mergedObj, idx);
            [~, order] = sort(d, 'descend');
            take = idx(order(1:popSize - size(newPop, 1)));
            newPop = [newPop; merged(take, :)]; %#ok<AGROW>
            newObj = [newObj; mergedObj(take, :)]; %#ok<AGROW>
            break
        end
    end
    pop = newPop; objs = newObj;
end

[fronts, ~] = ndsort(objs);
front = fronts{1};
T = table(objs(front, 1), objs(front, 2), 'VariableNames', {'f1', 'f2'});
mmlib_report('NSGA-II', sprintf('Pareto解个数=%d', numel(front)), ...
    sprintf('f1范围=%.3f~%.3f', min(objs(front, 1)), max(objs(front, 1))), ...
    sprintf('f2范围=%.3f~%.3f', min(objs(front, 2)), max(objs(front, 2))));
mmlib_savetable(here, 'nsga2', T);

% 图件由 Python 生成（技能规则：交付图必须用 Python 绘制，MATLAB 仅负责计算）

function pick = tournament(rank, crowd)
i = randi(numel(rank)); j = randi(numel(rank));
if rank(i) ~= rank(j)
    if rank(i) < rank(j), pick = i; else, pick = j; end
else
    if crowd(i) >= crowd(j), pick = i; else, pick = j; end
end
end

function [fronts, rank] = ndsort(objs)
n = size(objs, 1);
domBy = cell(n, 1); count = zeros(n, 1);
for p = 1:n
    for q = 1:n
        if p == q, continue; end
        if all(objs(p, :) <= objs(q, :)) && any(objs(p, :) < objs(q, :))
            domBy{p} = [domBy{p} q];
        elseif all(objs(q, :) <= objs(p, :)) && any(objs(q, :) < objs(p, :))
            count(p) = count(p) + 1;
        end
    end
end
fronts = cell(n, 1); levels = 0; current = find(count == 0)';
while ~isempty(current)
    levels = levels + 1;
    fronts{levels} = current;
    nxt = [];
    for p = current
        for q = domBy{p}
            count(q) = count(q) - 1;
            if count(q) == 0
                nxt = [nxt q]; %#ok<AGROW>
            end
        end
    end
    current = nxt;
end
fronts = fronts(1:levels);
rank = zeros(n, 1);
for k = 1:numel(fronts)
    rank(fronts{k}) = k;
end
end

function d = crowdDist(objs, front)
m = numel(front); d = zeros(m, 1);
if m <= 2
    d(:) = inf; return
end
vals = objs(front, :);
for k = 1:size(vals, 2)
    [~, order] = sort(vals(:, k));
    d(order(1)) = inf; d(order(end)) = inf;
    span = vals(order(end), k) - vals(order(1), k);
    if span > 0
        for i = 2:m - 1
            d(order(i)) = d(order(i)) + (vals(order(i + 1), k) - vals(order(i - 1), k)) / span;
        end
    end
end
end

File: skills/cumcm-code-writer/references/algorithm_library/01_优化与启发式/nsga2/python/main.py
"""NSGA-II · 双目标优化（Python 最小可运行模板）

模型：min f1 = Σ x_i^2，min f2 = Σ (x_i - 2)^2，x_i ∈ [0, 2]，n = 10
步骤：快速非支配排序 → 拥挤度距离 → 锦标赛选择 → SBX 交叉 → 多项式变异。
接口：solve(pop_size, generations, seed) -> (Pareto 前沿, 目标值)；运行：python main.py
"""
from __future__ import annotations

import sys
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "_common" / "python"))
from mmlib import demo, report, save_fig, save_table, set_style  # noqa: E402

DIMS, LOW, HIGH = 10, 0.0, 2.0


def build_demo() -> dict:
    return {"dims": DIMS, "low": LOW, "high": HIGH}


def objectives(pop):
    return np.column_stack([(pop ** 2).sum(1), ((pop - 2.0) ** 2).sum(1)])


def dominates(a, b) -> bool:
    return bool(np.all(a <= b) and np.any(a < b))


def fast_nondominated_sort(objs) -> list[list[int]]:
    n = len(objs)
    dominated_by = [[] for _ in range(n)]
    count = np.zeros(n, int)
    fronts = [[]]
    for p in range(n):
        for q in range(n):
            if p == q:
                continue
            if dominates(objs[p], objs[q]):
                dominated_by[p].append(q)
            elif dominates(objs[q], objs[p]):
                count[p] += 1
        if count[p] == 0:
            fronts[0].append(p)
    i = 0
    while fronts[i]:
        nxt: list[int] = []
        for p in fronts[i]:
            for q in dominated_by[p]:
                count[q] -= 1
                if count[q] == 0:
                    nxt.append(q)
        i += 1
        fronts.append(nxt)
    return [front for front in fronts if front]


def crowding_distance(objs, front: list[int]) -> np.ndarray:
    distance = np.zeros(len(front))
    if len(front) <= 2:
        return np.full(len(front), np.inf)
    values = objs[front]
    for k in range(values.shape[1]):
        order = np.argsort(values[:, k])
        distance[order[0]] = distance[order[-1]] = np.inf
        span = values[order[-1], k] - values[order[0], k]
        if span > 0:
            for i in range(1, len(front) - 1):
                distance[order[i]] += (values[order[i + 1], k] - values[order[i - 1], k]) / span
    return distance


def solve(pop_size: int = 40, generations: int = 30, seed: int = 42):
    rng = np.random.default_rng(seed)
    pop = rng.uniform(LOW, HIGH, size=(pop_size, DIMS))
    objs = objectives(pop)
    for _ in range(generations):
        fronts = fast_nondominated_sort(objs)
        rank = np.zeros(pop_size, int)
        crowd = np.zeros(pop_size)
        for level, front in enumerate(fronts):
            rank[front] = level
            crowd[front] = crowding_distance(objs, front)

        def tournament() -> int:
            i, j = rng.integers(0, pop_size, 2)
            if rank[i] != rank[j]:
                return int(i if rank[i] < rank[j] else j)
            return int(i if crowd[i] >= crowd[j] else j)

        children = np.empty_like(pop)
        for k in range(0, pop_size, 2):
            p1, p2 = pop[tournament()].copy(), pop[tournament()].copy()
            u = rng.random(DIMS)
            beta = np.where(u <= 0.5, (2 * u) ** (1 / 3), (1 / (2 * (1 - u))) ** (1 / 3))
            c1 = 0.5 * ((1 + beta) * p1 + (1 - beta) * p2)
            c2 = 0.5 * ((1 - beta) * p1 + (1 + beta) * p2)
            for child in (c1, c2):
                mask = rng.random(DIMS) < 0.1
                child[mask] += rng.normal(0, 0.1, mask.sum())
            children[k] = np.clip(c1, LOW, HIGH)
            if k + 1 < pop_size:
                children[k + 1] = np.clip(c2, LOW, HIGH)

        merged = np.vstack([pop, children])
        merged_objs = objectives(merged)
        new_pop, new_objs = [], []
        for front in fast_nondominated_sort(merged_objs):
            if len(new_pop) + len(front) <= pop_size:
                new_pop += [merged[i] for i in front]
                new_objs += [merged_objs[i] for i in front]
            else:
                distance = crowding_distance(merged_objs, front)
                order = np.argsort(-distance)
                for i in order[: pop_size - len(new_pop)]:
                    new_pop.append(merged[front[i]])
                    new_objs.append(merged_objs[front[i]])
                break
        pop, objs = np.array(new_pop), np.array(new_objs)

    front = fast_nondominated_sort(objs)[0]
    return pop[front], objs[front]


if __name__ == "__main__":
    demo(__file__, build_demo)
    pop, objs = solve()
    order = np.argsort(objs[:, 0])
    report("NSGA-II", Pareto解个数=len(pop), f1范围=f"{objs[:, 0].min():.3f}~{objs[:, 0].max():.3f}",
           f2范围=f"{objs[:, 1].min():.3f}~{objs[:, 1].max():.3f}")
    save_table(__file__, "pareto前沿", {"f1": np.round(objs[order, 0], 6), "f2": np.round(objs[order, 1], 6)})
    save_table(__file__, "pareto解", {f"x{i + 1}": np.round(pop[order, i], 6) for i in range(DIMS)})
    plt = set_style()
    fig, ax = plt.subplots()
    ax.scatter(objs[:, 0], objs[:, 1], color="#0072B2", s=22)
    ax.set_xlabel("$f_1$：Σ x²")
    ax.set_ylabel("$f_2$：Σ (x-2)²")
    ax.set_title("NSGA-II Pareto 前沿")
    save_fig(__file__, fig, "nsga2")
